### ISSUE TITLE
Accuracy for BERT

### ISSUE THREAD

**Issue by sarahbashir** [mikeizbicki/cmc-csci181#45]
I'm getting really bad accuracies (never higher than 40% for bert@20 and 15% for bert@1). I've tried many different hyperparameters and can't get higher than these results.. could it still be something with my hyperparameters (do I need to warmstart?) or is my code going wrong somewhere? 

```

def str_to_tensor(ss,input_length=None):
    '''
    Converts a list of strings into a tensor of shape <max_length, len(ss), len(vocabulary)>.
    This is used to convert text into a form suitable for input into a RNN/CNN.
    '''
    for i,line in enumerate(ss):
        tokens = tokenizer.tokenize(line)
        #print("tokens=",tokens)
    #crash
    # generates 1-hot encodings of the lines
    max_length = args.input_length
    if input_length:
        max_length = input_length
    encodings = []
    #tensor = torch.zeros(max_length, len(ss), len(vocabulary)).to(device)
    for i,line in enumerate(ss):
        encoding = tokenizer.encode_plus(
            line,
            max_length = max_length,
            pad_to_max_length = True,
            return_tensors = 'pt',
            )
        encodings.append(encoding)

    input_ids = torch.cat([encoding['input_ids'] for encoding in encodings ],dim=0)
    #print(input_ids.shape)

    return input_ids
```

```

class BertFineTuning(nn.Module):
    def __init__(self):
        super().__init__()
        #self.bert = transformers.BertModel.from_pretrained(checkpoint_name)
        #change 768 to? should be embedding size?
        self.fc = nn.Linear(768,len(all_categories))

    def forward(self,x):
        #last_layer,embedding = self.bert(x)
        last_layer,embedding = bert(x)
        embedding = torch.mean(last_layer,dim=1)
        out = self.fc(embedding)
        return out, torch.zeros(args.input_length,args.batch_size,76) #tensor filled with 0's same shape as out_nextchars

```

```

# perform training step
        output_class,output_nextchars = model(input_tensor)
        loss_class = criterion(output_class, category_tensor)

        if args.conditional_model or args.disable_categories:
            loss = loss_class
        else:
            loss = loss_class 
        loss.backward()

```

These are some of the hyper parameters I've tried:

`python3 project_part2.py --train --data=corona.multilang100/corona.multilang100.jsonl.gz --data_format=headlines --model=bert --hidden_layer_size=128 --num_layers=20  --optimizer=sgd --learning_rate=1e-3 --batch_size=4 --input_length=64`

**Comment by mikeizbicki** [comment]
`--hidden_layer_size` and `--num_layers` are not used by the bert model, and so changing those won't have an effect on your model.  Your other hyperparameters seem reasonable.

If your getting accuracies of 15%, that means your model is actually learning something, so my guess is that you just need to let it run for longer to get better results.  In particular, this model will take several days on a laptop to train fully to convergence.  You're not required to train to convergence though.

**Comment by sarahbashir** [comment]
As soon as the step reaches about 4k the model peaks and accuracy is only decreasing after that point... What does this indicate about the hyperparameters?

**Comment by mikeizbicki** [comment]
Hmm.... that's odd.  I'd need to see the tensorboard output to say for sure.  Can you paste a link to tensorboard.dev?

**Comment by sarahbashir** [comment]
Sure! https://tensorboard.dev/experiment/5YGTwpN8T7yv4t8vWX24hg/#scalars&_smoothingWeight=0.994

**Comment by mikeizbicki** [comment]
Two things jump out to me:

1. Your `grad_norm` is growing really large, and this can result in the accuracy decreasing after it gets too big.  Applying gradient clipping may help keep it small.

2. Your loss value is increasing.  It is rare to have your loss value go up and your accuracy go up as well; usually they are inversely correlated.  Typically I have only seen this happen before using Adam, and never with SGD.  I suspect that if you either: reduce your learning rate, or apply gradient clipping (as with the problem above), then you will see this problem go away.

**Comment by sarahbashir** [comment]
Applying gradient clipping fixed it, thank you!!

**Comment by mikeizbicki** [comment]
Reopenning to make it easier for others to find.
